File: January2022/C++/Linked_List_Cycle_II.cpp
created by @Pranav108
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        auto p = head;
        auto q = head;
        bool notFound = true;
        while(p&&q&&q->next&&notFound){
            p = p->next;
            q = q->next->next;
            if(p == q)
                notFound = false;
        }
        if(notFound) return NULL;
        p = head;
        while(p != q){
            p = p->next;
            q = q->next;
        }
        return p;
    }
};

File: January2022/Java/LinkedListCycleII.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
	public ListNode detectCycle(ListNode head) {
		if(head == null){
			return head;
		}
		// fast pointer slow pointer technique
		ListNode fast = head;
		ListNode slow = head;
		boolean start = true;
		while(fast!=slow  || start){
			start = false;
			if(slow!=null)
				slow  = slow.next;
			if(fast!=null)
				fast = fast.next;
			if(fast!=null) {
				fast = fast.next;
			}
		}

		// System.out.println(fast.val);

		fast = head;
		while(slow!=fast){
			if(fast!=null)
				fast = fast.next;
			if(slow!=null)
				slow = slow.next;
		}
		return fast;

	}
}